monotonic.go

```diff
@@ -0,0 +1,59 @@
+package ulid
+
+import (
+	"errors"
+	"io"
+)
+
+// Generator generates ULIDs.
+type Generator func(uint64) (ULID, error)
+
+// Monotonic returns a Generator of ULID that guarantees the ULIDs will be strictly
+// monotonically increasing.
+func Monotonic(entropy io.Reader) Generator {
```

Comment:
This breaks the consistency of the library APIs by introducing a new stateful `Generator` interface for this single use case and its accompanying `Monotonic` fabric implementation.

While not strictly equivalent to the solution here proposed, the underlying monotonicity problem seems to be solvable by encapsulating the state within the entropy source as an `io.Reader` which would have to guarantee consecutive calls would return strictly monotonic data, regardless of the value of the time component.

Reply:
Additionally, the behavior of such code under concurrent execution should be clearly documented.

Reply:
@tsenart 
> This breaks the consistency of the library APIs by introducing a new stateful Generator interface for this single use case and its accompanying Monotonic fabric implementation.

Agreed. I mean this current patch as a starting point, I'm aware it breaks the style of the library.

* * *
I'm not totally following what you mean by:
> While not strictly equivalent to the solution here proposed, the underlying monotonicity problem seems to be solvable by encapsulating the state within the entropy source as an io.Reader which would have to guarantee consecutive calls would return strictly monotonic data, regardless of the value of the time component.

Can you roughly sketch/pseudo-code your intent? 

As I understand it, making an `io.Reader` that guarantee something about consecutive calls means shifting the statefulness in that `io.Reader`. Also I'm not sure how putting this into the `io.Reader` would work, since the `io.Reader` would need to know something about the current timestamp. 

Unless you mean decoupling the monotonicity from time altogether and making an `io.Reader` that returns monotonically increasing random values? If that's the case, I feel this would skew the distribution of randomness given by the source of entropy, eventually leading to an entropy source that's basically a counter. Note that this is true when increasing the entropy within a single millisecond.

Reply:
> Unless you mean decoupling the monotonicity from time altogether and making an io.Reader that returns monotonically increasing random values? 

That was my train of thought.

> If that's the case, I feel this would skew the distribution of randomness given by the source of entropy, eventually leading to an entropy source that's basically a counter. Note that this is true when increasing the entropy within a single millisecond.

I see, you are right. Can you point me to the original discussion and rational for these monotonic requirements? I'd like to understand the context better.

> As I understand it, making an io.Reader that guarantee something about consecutive calls means shifting the statefulness in that io.Reader. Also I'm not sure how putting this into the io.Reader would work, since the io.Reader would need to know something about the current timestamp.

While a bit of hack, could we reserve the first bytes of the passed `p []byte` for the current timestamp in `ms`? That way we'd know for which timestamp the `io.Reader` would be generating entropy. What's your take on this rather dubious approach @peterbourgon? 🗡

Reply:
@tsenart the original issue was https://github.com/oklog/ulid/issues/6

Reply:
@tsenart Hacky indeed, I think this is worth a post to golang-nuts!